### PR TITLE
feat: added strategy with ape template

### PR DESCRIPTION
**Pull request by bufander**

### CODE REVIEW

**Review by jmonteer — CHANGES_REQUESTED** [review]
This strategy is liquid so it should be able to call `_freeFunds` from `withdraw`, that needs to be overriden to include it (so that the vault doesn't need anyone else freeing funds in advance).

Probably the same with _invest and deposit

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    # Use 1M
+    return 1_000_000 * 10 ** asset.decimals()
+
+
+@pytest.fixture(scope="session")
+def atoken():
+    return Contract(AASSET_ADDRESS)
+
+
+# TODO: deploying vault, as there is no vault yet on mainnet. To be deleted once vault v3 is deployed
```
[the remark this answers is not in the document]
Comment (bufander) [reply tests/conftest.py:43]: fair enough, although we need to figure out a better way of importing the VaultV3 code, as copy/pasting it on the repo is not good enough

**Comment by bufander** [comment]
@jmonteer I've re-written the strategy in a simpler way, just using minimum api vault needs. Please give it a look

**Review by jmonteer — COMMENTED** [review]
I would use this work to separate between BaseStrategy.sol (implementing all external functions) and Strategy.sol that inherits from it but implements only functions that are coing to be unique for 1 strategy

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    function balanceOf(address _address) public view returns (uint256) {
+        return balances[_address];
+    }
+
+    function deposit(uint256 assets, address receiver)
+        public
+        returns (uint256)
+    {
+        // transfer and invest
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol:71]: Done

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        return _convert_to_assets(shares);
+    }
+
+    function convertToShares(uint256 assets) public view returns (uint256) {
+        return _convert_to_shares(assets);
+    }
+
+    function pricePerShare() public view returns (uint256) {
+        return _convert_to_assets(10**IVault(vault).decimals());
+    }
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol:57]: Done

File: contracts/Strategy.sol [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            // We need to take from Aave enough to reach _amount
+            // Balance of
+            // We run with 'unchecked' as we are safe from underflow
+            unchecked {
+                _withdrawFromAave(
+                    Math.min(
+                        _amount - idle_amount,
+                        Math.min(
+                            balanceOfAToken(),
+                            IERC20(asset).balanceOf(aToken)
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol:130]: Good catch!

File: contracts/Strategy.sol [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        IProtocolDataProvider(0x057835Ad21a177dbdd3090bB1CAE03EaCF78Fc6d);
+
+    string public name;
+    address public immutable aToken;
+    address public immutable asset;
+    address public vault;
+    uint256 public totalSupply;
+
+    //    shares
+    mapping(address => uint256) private balances;
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol:27]: done

File: contracts/Strategy.sol [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+    function totalAssets() public view returns (uint256) {
+        return _totalAssets();
+    }
+
+    function balanceOf(address _address) public view returns (uint256) {
+        return balances[_address];
+    }
+
+    function deposit(uint256 assets, address receiver)
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol]: Moved

File: contracts/Strategy.sol [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                            IERC20(asset).balanceOf(aToken)
+                        )
+                    )
+                );
+            }
+            _amountFreed = balanceOfAsset();
+        }
+    }
+
+    function _withdraw(
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol:54]: Moved

File: contracts/Strategy.sol [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+    function _convert_to_assets(uint256 shares) public view returns (uint256) {
+        // if total_supply is 0, price_per_share is 1
+        if (totalSupply == 0) {
+            return shares;
+        }
+        return (shares * totalAssets()) / totalSupply;
+    }
+
+    function _convert_to_shares(uint256 assets) public view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol]: Moved

File: contracts/Strategy.sol [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    function withdraw(
+        uint256 amount,
+        address receiver,
+        address owner
+    ) public returns (uint256) {
+        require(amount <= maxWithdraw(owner), "withdraw more than max");
+        return _withdraw(amount, receiver, owner);
+    }
+
+    function _convert_to_assets(uint256 shares) public view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol]: Moved

File: contracts/Strategy.sol [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+    function convertToShares(uint256 assets) public view returns (uint256) {
+        return _convert_to_shares(assets);
+    }
+
+    function pricePerShare() public view returns (uint256) {
+        return _convert_to_assets(10**IVault(vault).decimals());
+    }
+
+    function totalAssets() public view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol]: Moved

**Comment by bufander** [comment]
@jmonteer  
Following your comments I've created the BaseStrategy. If it fits you I can update ape strategy mix with this new BaseStrategy

**Review by jmonteer — CHANGES_REQUESTED** [review]
last changes

File: contracts/BaseStrategy.sol [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    function withdraw(
+        uint256 amount,
+        address receiver,
+        address owner
+    ) public returns (uint256) {
+        require(msg.sender == owner, "not owner");
+        require(amount <= maxWithdraw(owner), "withdraw more than max");
+        return _withdraw(amount, receiver, owner);
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/BaseStrategy.sol]: Moved there

File: contracts/BaseStrategy.sol [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    function _withdraw(
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/BaseStrategy.sol:87]: I've taken all code from the method

File: contracts/Strategy.sol [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    }
+
+    function _withdraw(
+        uint256 amount,
+        address receiver,
+        address owner
+    ) internal override returns (uint256) {
+        uint256 amount_to_withdraw = _freeFunds(amount);
+        IERC20(asset).transfer(receiver, amount_to_withdraw);
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/Strategy.sol]: Done!

File: contracts/BaseStrategy.sol [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        require(msg.sender == vault && msg.sender == receiver, "not owner");
+
+        // transfer and invest
+        IERC20(asset).transferFrom(vault, address(this), assets);
+        _invest();
+        return assets;
+    }
+
+    function maxWithdraw(address owner) public view virtual returns (uint256) {
+        return _totalAssets();
```
[the remark this answers is not in the document]
Comment (bufander) [reply contracts/BaseStrategy.sol]: Ok, I implemented it that way so it´s more clear

**Pull request merged**